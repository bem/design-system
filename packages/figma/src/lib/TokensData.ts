export type TokenPrimitiveValue = string | number

export type TokenValue = {
  value: TokenPrimitiveValue
  description?: string
}

export type TokensGroup = { [key: string]: TokenPrimitiveValue | TokenValue | TokensGroup }

export const isTokenPrimitiveValue = (value: any): value is TokenPrimitiveValue => {
  return typeof value === 'string' || typeof value === 'number'
}

export const isTokenValue = (value: any): value is TokenValue => {
  return !isTokenPrimitiveValue(value) && 'value' in value
}

export const isTokensGroup = (value: any): value is TokensGroup => {
  return !(isTokenPrimitiveValue(value) || isTokenValue(value))
}

export class TokensData {
  tokens: TokensGroup = {}

  resolvedTokens: TokensGroup = {}

  constructor(public tokensJSON: string) {
    this.setTokens()
    this.setResolvedTokens()
  }

  private setTokens() {
    this.tokens = JSON.parse(this.tokensJSON)
  }

  private setResolvedTokens() {
    this.resolvedTokens = JSON.parse(this.tokensJSON, this.resolveAliases)
  }

  private resolveAliases = (key: string, value: TokenPrimitiveValue) => {
    if (typeof value === 'string' && TokensData.isAlias(value)) {
      // @todo: what about circularly references in aliases
      const nextValue = this.getTokenValueByAlias(value)
      if (typeof nextValue === 'string' && TokensData.isAlias(nextValue)) {
        return this.getTokenValueByAlias(nextValue)
      }
      return nextValue
    }
    return value
  }

  static isAlias(value: string) {
    return value.startsWith('{') && value.endsWith('}')
  }

  getTokenValueByAlias(value: string) {
    const parts = (value.slice(1, value.length - 1) as string).split('.')
    let val: any = this.tokens
    for (const part of parts) {
      if (typeof val[part] !== 'undefined') {
        val = val[part]
      } else {
        // @todo: error handling
        return value
      }
    }
    return val
  }
}
