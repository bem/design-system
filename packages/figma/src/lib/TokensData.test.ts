import { isTokenPrimitiveValue, isTokensGroup, isTokenValue, TokensData } from './TokensData'

describe('TokenData', () => {
  test('tokensJSON set', () => {
    expect(new TokensData(JSON.stringify('{}')).tokensJSON).toBeTruthy()
  })

  test('#getTokenValueByAlias return correct alias value', () => {
    const tokens = {
      colors: {
        Red: {
          '100': '#ff0000',
        },
        Action: {
          Focus: '{colors.Red.100}',
        },
      },
    }
    expect(new TokensData(JSON.stringify(tokens)).getTokenValueByAlias('{colors.Red.100}')).toBe('#ff0000')
  })

  test('static#isAlias', () => {
    expect(TokensData.isAlias('{colors.Red.100}')).toBeTruthy()
    expect(TokensData.isAlias('colors.Red.100')).toBeFalsy()
    expect(TokensData.isAlias('$colors.Red.100')).toBeFalsy()
  })

  test('#resolvedTokens get resolved tokens', () => {
    const tokens = {
      colors: {
        Red: {
          '100': '#ff0000',
        },
        Action: {
          Focus: '{colors.Red.100}',
        },
      },
    }
    expect(new TokensData(JSON.stringify(tokens)).resolvedTokens).toEqual({
      colors: {
        Red: {
          '100': '#ff0000',
        },
        Action: {
          Focus: '#ff0000',
        },
      },
    })
  })
  test('#resolvedTokens get resolved tokens alias-to-alias', () => {
    const tokens = {
      colors: {
        Red: {
          '50': '#ff0000',
          '100': '{colors.Red.50}',
        },
        Action: {
          Focus: '{colors.Red.100}',
        },
      },
    }
    expect(new TokensData(JSON.stringify(tokens)).resolvedTokens).toEqual({
      colors: {
        Red: {
          '50': '#ff0000',
          '100': '#ff0000',
        },
        Action: {
          Focus: '#ff0000',
        },
      },
    })
  })
})

describe('isTokenPrimitiveValue', () => {
  test('return correct value', () => {
    expect(isTokenPrimitiveValue('100px')).toBeTruthy()
    expect(isTokenPrimitiveValue(100)).toBeTruthy()
    expect(isTokenPrimitiveValue({ blah: {} })).toBeFalsy()
    expect(isTokenPrimitiveValue({ value: 'rgba(0 0 0)' })).toBeFalsy()
  })
})

describe('isTokenValue', () => {
  test('return correct value', () => {
    expect(isTokenValue({ value: 'rgba(0 0 0)' })).toBeTruthy()
    expect(isTokenValue('100px')).toBeFalsy()
    expect(isTokenValue(100)).toBeFalsy()
    expect(isTokenValue({ colors: {} })).toBeFalsy()
  })
})

describe('isTokenGroup', () => {
  test('return correct value', () => {
    expect(isTokensGroup({ colors: {} })).toBeTruthy()
    expect(isTokensGroup({ value: 'rgba(0 0 0)' })).toBeFalsy()
    expect(isTokensGroup('100px')).toBeFalsy()
    expect(isTokensGroup(100)).toBeFalsy()
  })
})
