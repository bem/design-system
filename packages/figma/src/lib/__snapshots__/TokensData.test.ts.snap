// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokensData #resolvedTokens get resolved tokens 1`] = `
Object {
  "colors": Object {
    "Action": Object {
      "Focus": "#ff0000",
    },
    "Red": Object {
      "100": "#ff0000",
    },
  },
}
`;
